File: typescript/packages/mechanisms/evm/test/unit/batch-settlement/delegatedAuth.test.ts
import { describe, it, expect } from "vitest";
import {
  DelegatedAuthIdentityConflictError,
  InMemoryDelegatedAuthStore,
} from "../../../src/batch-settlement/storage/delegatedAuth";

const NETWORK = "eip155:84532";
const CHANNEL_ID = "0xabc1230000000000000000000000000000000000000000000000000000000001";

describe("InMemoryDelegatedAuthStore", () => {
  it("binds a caller identity to a channel and returns a copy on get", async () => {
    const store = new InMemoryDelegatedAuthStore();
    await store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" });
    const row = await store.get(CHANNEL_ID, NETWORK);
    expect(row).toEqual({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" });
    row!.callerIdentity = "mutated";
    expect((await store.get(CHANNEL_ID, NETWORK))?.callerIdentity).toBe("tenant-a");
  });

  it("treats a repeat bind with the same identity as idempotent", async () => {
    const store = new InMemoryDelegatedAuthStore();
    await store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" });
    await expect(
      store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" }),
    ).resolves.toBeUndefined();
  });

  it("rejects a second identity for the same channel", async () => {
    const store = new InMemoryDelegatedAuthStore();
    await store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" });
    await expect(
      store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-b" }),
    ).rejects.toBeInstanceOf(DelegatedAuthIdentityConflictError);
  });

  it("delete removes the binding so a new identity can bind", async () => {
    const store = new InMemoryDelegatedAuthStore();
    await store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-a" });
    await store.delete(CHANNEL_ID, NETWORK);
    expect(await store.get(CHANNEL_ID, NETWORK)).toBeUndefined();
    await store.bind({ channelId: CHANNEL_ID, network: NETWORK, callerIdentity: "tenant-b" });
    expect((await store.get(CHANNEL_ID, NETWORK))?.callerIdentity).toBe("tenant-b");
  });
});

File: go/mechanisms/evm/batch-settlement/storage/delegated_auth_test.go
package storage

import (
	"context"
	"errors"
	"testing"
)

const (
	delegatedNetwork   = "eip155:84532"
	delegatedChannelId = "0xabc1230000000000000000000000000000000000000000000000000000000001"
)

func TestInMemoryDelegatedAuthStore_BindAndGetCopy(t *testing.T) {
	store := NewInMemoryDelegatedAuthStore()
	inserted, err := store.Bind(context.Background(), DelegatedAuthBinding{
		ChannelId:      delegatedChannelId,
		Network:        delegatedNetwork,
		CallerIdentity: "tenant-a",
	})
	if err != nil || !inserted {
		t.Fatalf("Bind: inserted=%v err=%v", inserted, err)
	}
	row, err := store.Get(context.Background(), delegatedChannelId, delegatedNetwork)
	if err != nil {
		t.Fatalf("Get: %v", err)
	}
	if row == nil || row.CallerIdentity != "tenant-a" || row.ChannelId != delegatedChannelId || row.Network != delegatedNetwork {
		t.Fatalf("row = %+v", row)
	}
	row.CallerIdentity = "mutated"
	again, err := store.Get(context.Background(), delegatedChannelId, delegatedNetwork)
	if err != nil {
		t.Fatalf("Get after mutate: %v", err)
	}
	if again.CallerIdentity != "tenant-a" {
		t.Fatalf("stored row mutated: %q", again.CallerIdentity)
	}
}

func TestInMemoryDelegatedAuthStore_RepeatBindIsIdempotent(t *testing.T) {
	store := NewInMemoryDelegatedAuthStore()
	binding := DelegatedAuthBinding{ChannelId: delegatedChannelId, Network: delegatedNetwork, CallerIdentity: "tenant-a"}
	inserted, err := store.Bind(context.Background(), binding)
	if err != nil || !inserted {
		t.Fatalf("Bind: inserted=%v err=%v", inserted, err)
	}
	inserted, err = store.Bind(context.Background(), binding)
	if err != nil || inserted {
		t.Fatalf("repeat Bind: inserted=%v err=%v", inserted, err)
	}
}

func TestInMemoryDelegatedAuthStore_RejectsSecondIdentity(t *testing.T) {
	store := NewInMemoryDelegatedAuthStore()
	if _, err := store.Bind(context.Background(), DelegatedAuthBinding{ChannelId: delegatedChannelId, Network: delegatedNetwork, CallerIdentity: "tenant-a"}); err != nil {
		t.Fatalf("Bind: %v", err)
	}
	_, err := store.Bind(context.Background(), DelegatedAuthBinding{ChannelId: delegatedChannelId, Network: delegatedNetwork, CallerIdentity: "tenant-b"})
	var conflict *DelegatedAuthIdentityConflictError
	if !errors.As(err, &conflict) {
		t.Fatalf("err = %v, want DelegatedAuthIdentityConflictError", err)
	}
}

func TestInMemoryDelegatedAuthStore_DeleteAllowsRebind(t *testing.T) {
	store := NewInMemoryDelegatedAuthStore()
	if _, err := store.Bind(context.Background(), DelegatedAuthBinding{ChannelId: delegatedChannelId, Network: delegatedNetwork, CallerIdentity: "tenant-a"}); err != nil {
		t.Fatalf("Bind: %v", err)
	}
	if err := store.Delete(context.Background(), delegatedChannelId, delegatedNetwork); err != nil {
		t.Fatalf("Delete: %v", err)
	}
	row, err := store.Get(context.Background(), delegatedChannelId, delegatedNetwork)
	if err != nil || row != nil {
		t.Fatalf("Get after delete: row=%+v err=%v", row, err)
	}
	if _, err := store.Bind(context.Background(), DelegatedAuthBinding{ChannelId: delegatedChannelId, Network: delegatedNetwork, CallerIdentity: "tenant-b"}); err != nil {
		t.Fatalf("rebind: %v", err)
	}
	row, err = store.Get(context.Background(), delegatedChannelId, delegatedNetwork)
	if err != nil || row == nil || row.CallerIdentity != "tenant-b" {
		t.Fatalf("row after rebind = %+v err=%v", row, err)
	}
}
